### ISSUE TITLE
Parsing Conflict: split=val Default Argument vs. Boolean val

### ISSUE THREAD

**Issue by Sanjib-ac** [ultralytics/ultralytics#21550]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

What does the argument **split=val** mean? There is already a boolean argument named **val**, which causes a conflict when using the parser. If possible, please consider renaming one of them to **validation** to avoid this issue. Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Sanjib-ac, thank you for bringing this up and for using Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed explanations on arguments and configuration, as you may find additional helpful context there.

This is an automated response 🤖. An Ultralytics engineer will review your question and assist you further soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the parsing conflict—this will help us understand and address your concern more effectively.

If your question is related to custom training or usage, sharing specific usage examples or command-line calls can also help us support you better.

Join the Ultralytics community for real-time support and discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Run YOLO in any verified environment (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your feedback!

**Comment by RizwanMunawar** [comment]
@Sanjib-ac Thanks for pointing out the `split` argument. It works differently from the `val` argument used for model validation, for example, `yolo val data=coco.yaml`.

The `split` argument is used to specify **which dataset subfolder** to use for validation. For example, if your dataset has `train`, `val`, and `test` directories, the default is `split=val`, meaning the `val` folder will be used for validation. If you want to validate on the `test` set instead, you can simply use `split=test`, and the model will run validation on that subset.

Thanks.

**Comment by Y-T-G** [comment]
It should be a string value for `split`, `split="val"` which means run validation on that split. In CLI mode, `split=val` works because `val` is treated as string even without quotes.

There shouldn't be any conflict between that and the argument `val`.

**Comment by Sanjib-ac** [comment]
Thanks for your response. Here is the problem: I am parsing like this: 

`parser.add_argument("--val", action="store_true", default=False, help="Enables validation during training")`     and `parser.add_argument("--split", type=str, default=None)` 

that makes val =False below:

 ` ..., val=False, split=None, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=False, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=None, workspace=None, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None,` 

but still getting the detailed SegmentMetrics below:

 

```
Training completed successfully!
Results: ultralytics.utils.metrics.SegmentMetrics object with attributes:
ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x000001CA98161540>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)', 'Precision-Recall(M)', 'F1-Confidence(M)', 'Precision-Confidence(M)', 'Recall-Confidence(M)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.048048,    0.049049,  
```

I think that parser is not working well with the same naming conflict.

Thanks!

**Comment by glenn-jocher** [comment]
Thanks for the detailed example! You've identified an interesting issue. The two arguments serve different purposes but I can see how the similar naming could cause confusion:

- `val` (boolean): Controls whether validation runs **during training** ([reference](https://docs.ultralytics.com/usage/cfg/#training-settings))
- `split` (string): Specifies **which dataset split** to use when validation does run ([reference](https://docs.ultralytics.com/usage/cfg/#validation-settings))

However, you're correct that something seems off - if you set `val=False`, you shouldn't be getting validation metrics output after training. This suggests validation is still running despite your setting, which could indicate a bug.

Could you share the exact command or code you're using to start training? This would help determine if there's an issue with how the `val=False` parameter is being processed or if validation is being triggered through another mechanism.

**Comment by Sanjib-ac** [comment]
Below : 
```
def parse_args():    
    parser = argparse.ArgumentParser(description="Train a yolo model")
    parser.add_argument("--task", type=str, choices=["detect", "segment", "classify"],
                        default="detect", help="Task type: detect (default), segment, or classify")

    # Core
    parser.add_argument("--model", type=str, required=True,
                        help="weights or checkpoint(e.g., model.pt, or .yaml)")
    parser.add_argument("--data", type=str, required=True,
                        help="Dataset config YAML (train/val paths + class names)")
    parser.add_argument("--val", action="store_true", default=False, help="Enables validation during training")
```

```
def main():
      args = parse_args()
     train_kwargs = {
            "task": args.task,
            "model": args.model,
            "data": args.data,            
            "optimizer": args.optimizer,
            "patience": args.patience,
            "verbose": args.verbose,
            "val": args.val,
            "split": args.split,
            "plots": args.plots
        }
     model = YOLO(args.model)
     results = model.train(**train_kwargs)    
```
      Thanks!

**Comment by glenn-jocher** [comment]
Thanks for sharing the code! The issue you're experiencing where validation metrics appear despite `val=False` is normal. The final validation still runs even with `val=False`.

**Comment by Sanjib-ac** [comment]
Debug info: (CLI) `Arguments:Namespace(task='segment', model='.\\yolo11s-seg.pt', data='.\\dataSet_test_seg.yaml', pretrained=None, epochs=2, batch=4, imgsz=640, device='0', val=False, project='train', name='test', exist_ok=False, workers=8, rect=False, single_cls=False, multi_scale=False, amp=False, optimizer='auto', patience=50, verbose=False, shear=0.0, degrees=0.0, bgr=0.0, mixup=0.0, plots=False, split=None)`

validation still runs.

**Comment by glenn-jocher** [comment]
Thanks for the debug info! The `val=False` parameter only disables validation **during training epochs**. YOLO always runs a final validation after training completes, regardless of the `val` setting, which is why you're seeing the SegmentMetrics output.

To completely disable all validation including the final one, you would need to set `data=None` or ensure there's no validation split in your dataset configuration. However, this isn't typically recommended as you lose important model performance insights.

The `split` and `val` arguments don't actually conflict in practice - they serve different purposes and are processed separately by the argument parser.

**Comment by Sanjib-ac** [comment]
Okay, thanks!

**Comment by glenn-jocher** [comment]
You're welcome! Feel free to reach out if you have any other questions about YOLO training or validation settings.
